File: camerax_android/lib/src/core/resolution_selector/resolution_selector.dart
import 'package:camerax_android/src/jni.dart' as jni;
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class MyResolutionSelector implements ResolutionSelector {
  final jni.ResolutionSelector jniValue;

  MyResolutionSelector.jni(this.jniValue);

  factory MyResolutionSelector({
    AspectRatioStrategy? aspectRatioStrategy,
    ResolutionMode? allowedResolutionMode,
    ResolutionFilter? resolutionFilter,
    ResolutionStrategy? resolutionStrategy,
  }) {
    final jniBuilder = jni.ResolutionSelector_Builder();
    if (aspectRatioStrategy != null) {
      jniBuilder.setAspectRatioStrategy(aspectRatioStrategy.jniValue);
    }
    if (allowedResolutionMode != null) {
      jniBuilder.setAllowedResolutionMode(allowedResolutionMode.jniValue);
    }
    if (resolutionFilter != null) {
      jniBuilder.setResolutionFilter(resolutionFilter.jniValue);
    }
    if (resolutionStrategy != null) {
      jniBuilder.setResolutionStrategy(resolutionStrategy.jniValue);
    }
    final jniResolutionSelector = jniBuilder.build();
    return MyResolutionSelector.jni(jniResolutionSelector);
  }

  @override
  AspectRatioStrategy get aspectRatioStrategy {
    final strategy = jniValue.getAspectRatioStrategy();
    return strategy.dartValue;
  }

  @override
  ResolutionMode get allowedResolutionMode {
    final mode = jniValue.getAllowedResolutionMode();
    return mode.dartResolutionMode;
  }

  @override
  ResolutionFilter? get resolutionFilter {
    final filter = jniValue.getResolutionFilter();
    if (filter.isNull) {
      return null;
    }
    return filter.dartValue;
  }

  @override
  ResolutionStrategy? get resolutionStrategy {
    final strategy = jniValue.getResolutionStrategy();
    if (strategy.isNull) {
      return null;
    }
    return strategy.dartValue;
  }
}

File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/core/resolutionselector/ResolutionSelector.kt
package dev.hebei.camerax_android.legacy.core.resolutionselector

import dev.hebei.camerax_android.legacy.CameraXRegistrar
import dev.hebei.camerax_android.legacy.PigeonApiResolutionSelector
import dev.hebei.camerax_android.legacy.ResolutionMode

class ResolutionSelector(registrar: CameraXRegistrar) : PigeonApiResolutionSelector(registrar) {
    override fun pigeon_defaultConstructor(
        allowedResolutionMode: ResolutionMode,
        aspectRatioStrategy: androidx.camera.core.resolutionselector.AspectRatioStrategy,
        resolutionFilter: androidx.camera.core.resolutionselector.ResolutionFilter?,
        resolutionStrategy: androidx.camera.core.resolutionselector.ResolutionStrategy?
    ): androidx.camera.core.resolutionselector.ResolutionSelector {
        val builder = androidx.camera.core.resolutionselector.ResolutionSelector.Builder()
            .setAllowedResolutionMode(allowedResolutionMode.obj).setAspectRatioStrategy(aspectRatioStrategy)
        if (resolutionFilter != null) {
            builder.setResolutionFilter(resolutionFilter)
        }
        if (resolutionStrategy != null) {
            builder.setResolutionStrategy(resolutionStrategy)
        }
        return builder.build()
    }

    override fun allowedResolutionMode(pigeon_instance: androidx.camera.core.resolutionselector.ResolutionSelector): ResolutionMode {
        return pigeon_instance.allowedResolutionMode.resolutionModeArgs
    }

    override fun aspectRatioStrategy(pigeon_instance: androidx.camera.core.resolutionselector.ResolutionSelector): androidx.camera.core.resolutionselector.AspectRatioStrategy {
        return pigeon_instance.aspectRatioStrategy
    }

    override fun resolutionFilter(pigeon_instance: androidx.camera.core.resolutionselector.ResolutionSelector): androidx.camera.core.resolutionselector.ResolutionFilter? {
        return pigeon_instance.resolutionFilter
    }

    override fun resolutionStrategy(pigeon_instance: androidx.camera.core.resolutionselector.ResolutionSelector): androidx.camera.core.resolutionselector.ResolutionStrategy? {
        return pigeon_instance.resolutionStrategy
    }
}

val ResolutionMode.obj
    get() = when (this) {
        ResolutionMode.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION -> androidx.camera.core.resolutionselector.ResolutionSelector.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION
        ResolutionMode.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE -> androidx.camera.core.resolutionselector.ResolutionSelector.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE
    }

val Int.resolutionModeArgs
    get() = when (this) {
        androidx.camera.core.resolutionselector.ResolutionSelector.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION -> ResolutionMode.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION
        androidx.camera.core.resolutionselector.ResolutionSelector.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE -> ResolutionMode.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE
        else -> throw IllegalArgumentException()
    }